### DISCUSSION TITLE
73

### DISCUSSION THREAD

**Discussion by jojoldu** [jojoldu/blog-comments#762] [Comments]
http://jojoldu.tistory.com/73

**Comment by jojoldu** [comment]
작성자: 과객

okky란 사이트를 통해 들어왔습니다. 
열심히 하시는게 보기 좋네요. 블랙잭에 대한 의견 남겨봅니다.

블랙잭은 게임, 덱, 사용자, 딜러로 구성되어지죠. 이를 들여다 보면..

1. 덱 - 게임 객체의 상태.
54장의 카드생성, 셔플, 한 장꺼내는 기능을 가져야 하니
list comprehension 4-5줄로 게임 클래스 내에 배열로 구현하면 됩니다.

2. 게이머 - 사용자와 딜러
게이머는 카드와 상태(nil, burst, stay)를 두 가지를 갖습니다.
카드획득, 초이스(딜러여부에 따라 자동hit, 수동hit) 두 메소드를 구현하면 됩니다.

3. 게임 - gamers, deck을 갖는 클래스
초기화(덱 생성, 게이머 생성, 카드 2장 할당)
플레이(gamers.each { gamer.choice until 게이머.상태가 burst 또는 stay
저지(사용자의 상태와 카드합으로 승자 출력)

Gamer클래스를 상속하는 User와 Dealer를 분리 구현해도 되는데
둘을 구분하는 초이스 내용이 크지 않으니 하나로 합해서 구현해도 괜찮겠죠.
결과적으로 Game(덱) -> Gamer 또는 Game -> Gamer <|- User, Dealer 가 됩니다.
물리적으로는 2개 파일 소스는 언어에 따라 다르지만 대략 60~80라인 정도로 구현되겠네요.

**Comment by jojoldu** [comment]
작성자: 과객

펜더라는 분의 스칼라 소스를 잠시 봤습니다.
블랙잭이 물리적으로 29개 소스파일로 구성될 만큼 복잡한지를 생각해 봐야합니다.
개념과 로직의 과도한 분리는 지식의 파편화로 이어지기에 적당히 분리하는게 좋습니다.

패키지 제품 설계는 대체로 분리에 집중되는 경향이 있고
복잡한 도메인의 설계는 책임의 통합과 통합된 롤사이의 협력을 정의하는 것에 집중합니다.

문제가 고립되는지 협업사이즈가 얼마나 되는지에 따라 양상이 달라지는건데요.
둘 모두를 경험하며 균형감을 갖는게 중요하다고 생각됩니다.

정답이 아닌 지나가는 과객의 경험과 생각이니 가볍게 받아들여주시면 좋겠습니다. 건승하시길~!

**Comment by jojoldu** [comment]
작성자: fender

과객// 우선 의견 주셔서 감사드립니다.

설계에 있어서 절대적인 정답은 없다는 전제로 이야기하자면, 개인적으로 코드의 복잡성은 유형의 숫자나 라인수에 비례하지 않는다고 생각합니다.

물론 블랙잭 콘솔 게임 정도라면야 마음만 먹으면 메서드 몇 개에 다 구현해도 무방하겠습니다만, 애초에 블랙잭이라는 도메인을 고른 건 모델링 연습을 위한 것이니 만큼  최대한 개념의 역할이 겹치지 않고 확장할 수 있는 형태로 만드는데 집중했습니다.

게이머와 딜러를 구분하는 것을 조건문 플래그로 할지, 아니면 상속 관계의 유형으로 표현할지, 또 카드 덱을 그냥 문자와 숫자의 배열로 표현할지, 혹은 각각의 특화된 자료구조로 표현할 지는 역시 각각의 장단점이 존재하는 접근 방법일 것입니다.

하지만 당초 프로젝트의 의의가 콘솔 기반 블랙잭 게임을 가장 간결한 코드로 구현한다 같은 것이 아닌, 최대한 객체지향적인 모델링 방법을 적용해서 풀어내는 것임을 감안하면 기능에 비해 클래스 갯수가 많아지는 건 어느 정도 감안해야 하지 않나 생각합니다.

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

과객, fender 
앗 ㅠㅠ 두분 고수님들의 좋은말씀 감사드립니다!
과객님이 말씀하신 지식의 파편화에 대한 주의에 대한 이야기 정말 감사드립니다.
fender님이 얘기해주신 내용은 제가 처음 의도했던 목적을 다시 상시시키게 된것 같습니다. 감사드립니다!
이 다음의 내용은 좀더 고민을 해서 코드를 작성하겠습니다!

**Comment by dpudpu** [comment]
좋은 글 감사합니다.
항상 많이 배우고 있습니다.
